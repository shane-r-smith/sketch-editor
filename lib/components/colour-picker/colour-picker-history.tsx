import type { FunctionComponent } from "react";
import type { ColourPickerHistoryProps } from "./colour-picker.types";

import Grid from "@mui/material/Grid";
import Box from "@mui/material/Box";

export const ColourPickerHistory: FunctionComponent<
  ColourPickerHistoryProps
> = ({ history, numHistoryToShow, onColourSelected }) => {
  const unique = Array.from(new Set(history));
  const historyItemsToShow =
    unique.length < numHistoryToShow
      ? [...unique, ...Array(numHistoryToShow - unique.length).fill(undefined)]
      : unique.slice(0, numHistoryToShow);

  return (
    <Grid container spacing={1}>
      {historyItemsToShow.map((colour, i) => {
        return (
          <Box
            key={`${colour}-colour-history-${i}`}
            sx={[
              {
                width: "1.75rem",
                height: "1.75rem",
                borderRadius: "0.25rem",
                backgroundColor: "grey.50",
              },
              colour && {
                cursor: "pointer",
                backgroundColor: colour,
                "&:hover": {
                  outlineWidth: 2,
                  outlineStyle: "solid",
                  outlineColor: "primary.main",
                },
              },
            ]}
            onClick={() => {
              if (!colour) {
                return;
              }

              onColourSelected(colour);
            }}
          />
        );
      })}
    </Grid>
  );
};

export default ColourPickerHistory;
