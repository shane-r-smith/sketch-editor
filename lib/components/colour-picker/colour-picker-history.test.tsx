import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ColourPickerHistory } from "./colour-picker-history";

describe("<ColourPickerHistory />", () => {
  const defaultProps = {
    history: ["#ff0000", "#00ff00", "#0000ff"],
    numHistoryToShow: 3,
    onColourSelected: vi.fn(),
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<ColourPickerHistory {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("shows unique colours up to numHistoryToShow", () => {
    const { container } = render(<ColourPickerHistory {...defaultProps} />);
    const boxes = container.children[0].children;
    expect(boxes.length).toBe(defaultProps.numHistoryToShow);
  });

  it("fills with undefined when not enough history", async () => {
    const user = userEvent.setup();
    const props = {
      ...defaultProps,
      history: ["#ff0000"],
      numHistoryToShow: 3,
    };

    const { container } = render(<ColourPickerHistory {...props} />);
    const boxes = container.children[0].children;
    expect(boxes.length).toBe(defaultProps.numHistoryToShow);

    // Only the first box should have a clickable colour
    await user.click(boxes[0]);
    expect(props.onColourSelected).toHaveBeenCalledWith("#ff0000");

    // Clicking empty boxes should not call onColourSelected
    await user.click(boxes[1]);
    await user.click(boxes[2]);
    expect(props.onColourSelected).toHaveBeenCalledTimes(1);
  });
});
