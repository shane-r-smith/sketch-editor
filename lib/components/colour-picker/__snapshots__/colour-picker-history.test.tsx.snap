// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ColourPickerHistory /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-1 css-1yvsbrk-MuiGrid-root"
  >
    <div
      class="MuiBox-root css-10zmmb7"
    />
    <div
      class="MuiBox-root css-1wy9ye4"
    />
    <div
      class="MuiBox-root css-1qu7dg4"
    />
  </div>
</div>
`;
